src/arena.c

```diff
@@ -1283,19 +1294,38 @@ arena_bin_malloc_hard(tsdn_t *tsdn, arena_t *arena, bin_t *bin,
 	return arena_slab_reg_alloc(slab, bin_info);
 }
 
+/* Choose a bin shard and return the locked bin. */
+bin_t *
+arena_bin_choose_lock(tsdn_t *tsdn, arena_t *arena, szind_t binind,
+    unsigned *binshard) {
+	bin_t *bin;
+	if (tsdn_null(tsdn) || tsd_arena_get(tsdn_tsd(tsdn)) == NULL) {
+		*binshard = 0;
+	} else {
+		*binshard = tsd_binshard_get(tsdn_tsd(tsdn)) %
```

Comment:
The div instruction accounts for ~8% of tcache_fill_small. I'll see if our `div_compute` function can help here.

Reply:
Why do we need a % here at all?  Doesn't binshard_set already do the mod?

Reply:
We don't have an extent we can key off of here.

`div_compute` can't help, since it needs to know that the division is exact. However, we don't need to use `%` for range reduction; there are faster methods (e.g. https://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/ ).

Reply:
yeah I ended up doing a precompute per thread instead. In the long run we might want to make this more dynamic, instead of the current simple round-robin. So I figure a per thread bin-binding should be useful anyway.